Setuptools 62.2+ and twine require non-empty `long_description`

### Context
Since the last few days, an error can be encountered when building some projects: PyBuilder fails when running `twine check` on a project that does not define a `description` on its `build.py`. 
Here is part of the build log:
```
[INFO]  Building binary distribution in c:\***\sample-project\target\dist\project
[INFO]  Running Twine check for generated artifacts
------------------------------------------------------------
BUILD FAILED - Error while executing Twine ['check', 'c:\\***\\sample-project\\target\\dist\\project\\dist\\sample-project-1.0.0.dev20220513160132.tar.gz', 
'c:\\***\\sample-project\\target\\dist\\project\\dist\\sample_project-1.0.0.dev20220513160132-py3-none-any.whl']. 
See c:\***\sample-project\target\reports\distutils\twine_check.log for full details:
        Checking c:\***\sample-project\target\dist\project\dist\
        sample_project-1.0.0.dev20220513160132-py3-none-any.whl: FAILED
        ERROR    `long_description` has syntax errors in markup and would not be       
                 rendered on PyPI.                                                     
                                                                                       
        WARNING  `long_description_content_type` missing. defaulting to `text/x-rst`.  
        Checking c:\***\sample-project\target\dist\project\dist\
        sample-project-1.0.0.dev20220513160132.tar.gz: PASSED with warnings
        WARNING  `long_description_content_type` missing. defaulting to `text/x-rst`.  
        WARNING  `long_description` missing.                                           
         (lib\site-packages\pybuilder\plugins\python\distutils_plugin.py:421)
```
Please note that the project not having a `description` did not result in any problem before.

The error started arising after the release of Setuptools `v62.2.0` on 2022-05-10; if the line:
https://github.com/pybuilder/pybuilder/blob/8731de99ac280d3a82cad795bf7af015359c80c2/src/main/python/pybuilder/plugins/python/distutils_plugin.py#L117
is changed into:
```python
    project.plugin_depends_on("setuptools", "<62.2.0")
```
Setuptools `v62.1.0` is installed and the error does not occur.

### Proposal
Besides adapting PyBuilder to any new or changed Setuptools functionality, I would suggest to parametrize the version requirements for all plugin dependencies.
This way, if a third-party module update breaks part of PyBuilder's functionality, a fix can be easily rolled out by reverting to a previous version of the dependency, or the issue prevented altogether by freezing these dependencies' version at a point in time when the build is stable.

Actually, reading this I'm not sure there is a problem!

```
BUILD FAILED - Error while executing Twine ['check', 'c:\\***\\sample-project\\target\\dist\\project\\dist\\sample-project-1.0.0.dev20220513160132.tar.gz', 
'c:\\***\\sample-project\\target\\dist\\project\\dist\\sample_project-1.0.0.dev20220513160132-py3-none-any.whl']. 
See c:\***\sample-project\target\reports\distutils\twine_check.log for full details:
        Checking c:\***\sample-project\target\dist\project\dist\
        sample_project-1.0.0.dev20220513160132-py3-none-any.whl: FAILED
        ERROR    `long_description` has syntax errors in markup and would not be       
                 rendered on PyPI.                                                     
                                                                                       
        WARNING  `long_description_content_type` missing. defaulting to `text/x-rst`.  
        Checking c:\***\sample-project\target\dist\project\dist\
        sample-project-1.0.0.dev20220513160132.tar.gz: PASSED with warnings
        WARNING  `long_description_content_type` missing. defaulting to `text/x-rst`.  
        WARNING  `long_description` missing.                                           
         (lib\site-packages\pybuilder\plugins\python\distutils_plugin.py:421)
```

What are your projects' `description` settings @hoxu @lormico ?

--- a/.github/workflows/pybuilder.yml
+++ b/.github/workflows/pybuilder.yml
@@ -19,8 +19,6 @@ jobs:
           - '3.8'
           - '3.7'
           - '3.6'
-          - 'pypy-3.7'
-          - 'pypy-3.8'
         with-venv:
           - 'true'
           - 'false'
@@ -47,7 +45,6 @@ jobs:
       matrix:
         os:
           - windows-latest
-          - macos-10.15
           - macos-11
         python-version:
           - '3.10'
@@ -55,8 +52,6 @@ jobs:
           - '3.8'
           - '3.7'
           - '3.6'
-          - 'pypy-3.7'
-          - 'pypy-3.8'
         with-venv:
           - 'true'
           - 'false'
@@ -85,10 +80,12 @@ jobs:
         os:
           - ubuntu-latest
           - windows-latest
-          - macos-10.15
           - macos-11
         python-version:
-          - '3.11.0-alpha.5'
+          - '3.11.0-beta.1'
+          - 'pypy-3.7'
+          - 'pypy-3.8'
+          - 'pypy-3.9'
         with-venv:
           - 'true'
           - 'false'

--- a/src/integrationtest/python/should_write_namespace_packages_tests.py
+++ b/src/integrationtest/python/should_write_namespace_packages_tests.py
@@ -95,7 +95,7 @@ if __name__ == '__main__':
         name = 'integration-test',
         version = '1.0.dev0',
         description = '',
-        long_description = '',
+        long_description = 'integration-test',
         long_description_content_type = None,
         classifiers = [
             'Development Status :: 3 - Alpha',

--- a/src/integrationtest/python/should_write_setup_file_tests.py
+++ b/src/integrationtest/python/should_write_setup_file_tests.py
@@ -103,7 +103,7 @@ if __name__ == '__main__':
         name = 'integration-test',
         version = '1.0.dev0',
         description = '',
-        long_description = '',
+        long_description = 'integration-test',
         long_description_content_type = None,
         classifiers = [
             'Development Status :: 3 - Alpha',

--- a/src/integrationtest/python/should_write_setup_file_with_requirements_tests.py
+++ b/src/integrationtest/python/should_write_setup_file_with_requirements_tests.py
@@ -95,7 +95,7 @@ if __name__ == '__main__':
         name = 'integration-test',
         version = '1.0.dev0',
         description = '',
-        long_description = '',
+        long_description = 'integration-test',
         long_description_content_type = None,
         classifiers = [
             'Development Status :: 3 - Alpha',

--- a/src/main/python/pybuilder/plugins/python/distutils_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/distutils_plugin.py
@@ -180,6 +180,15 @@ def set_description(project, logger, reactor):
                     project.get_property("distutils_description_overwrite")):
                 setattr(project, "description", description)
 
+    if (not hasattr(project, "description") or
+            not project.description):
+        if hasattr(project, "summary") and project.summary:
+            description = project.summary
+        else:
+            description = project.name
+
+        setattr(project, "description", description)
+
     warn = False
     if len(project.summary) >= 512:
         logger.warn("Project summary SHOULD be shorter than 512 characters per PEP-426")

--- a/src/main/python/pybuilder/plugins/python/vendorize_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/vendorize_plugin.py
@@ -36,7 +36,8 @@ from pybuilder.utils import as_list, makedirs, jp, np
 
 __author__ = "Arcadiy Ivanov"
 
-RE_FROM_IMPORT = re.compile(r"from\s+(\S+)\s+import\s+")
+_RE_FROM_IMPORT = re.compile(r"from\s+(\S+)\s+import\s+")
+_RE_DECODE_PY = re.compile(rb'^[ \t\f]*#.*?coding[:=][ \t]*([-_.a-zA-Z0-9]+)')
 
 use_plugin("python.core")
 
@@ -255,7 +256,7 @@ class ImportTransformer(NodeVisitor):
                     import_changes = _relpkg_from(self.pkg_parts, module)
                     import_stmt, offset_start, offset_end = self.extract_source(node)
 
-                    m = RE_FROM_IMPORT.match(import_stmt)
+                    m = _RE_FROM_IMPORT.match(import_stmt)
 
                     ts = self.transformed_source
                     ts_prefix = ts[:offset_start + m.start(1) + self.offset]
@@ -276,15 +277,19 @@ def _vendorize(vendorized_path, logger):
         for py_path in files:
             if not py_path.endswith(".py"):
                 continue
+
             py_path = np(jp(root, py_path))
-            with open(py_path, "rt") as source_file:
-                source = source_file.read()
+            with open(py_path, "rb") as source_file:
+                source_b = source_file.read()
+            source_encoding, source = _decode_py(source_b)
+
             parsed_ast = ast.parse(source, filename=py_path)
             it = ImportTransformer(py_path, source, vendorized_path, vendorized_packages, [])
             it.visit(parsed_ast)
+
             if source != it.transformed_source:
                 logger.debug("Vendorized %r", py_path)
-                with open(py_path, "wt") as source_file:
+                with open(py_path, "wt", encoding=source_encoding) as source_file:
                     source_file.write(it.transformed_source)
 
     return vendorized_packages
@@ -339,3 +344,20 @@ def _extract_source(source_lines, node):
     end_line_offset = source_lines[end_lineno]
 
     return start_line_offset + col_offset, end_line_offset + end_col_offset
+
+
+def _decode_py(source_b):
+    encoding = "utf-8"
+    if source_b.startswith(b'\xef\xbb\xbf'):
+        encoding = "utf-8"
+    else:
+        source_b_lines = source_b.splitlines()
+        for idx, line in enumerate(source_b_lines):
+            if idx > 1:
+                break
+            match = _RE_DECODE_PY.match(line)
+            if match:
+                encoding = match.group(1).decode("utf-8")
+                break
+
+    return encoding, source_b.decode(encoding, errors='strict')

